### PR TITLE
Enable RTL support for all inputs and textareas

### PR DESCRIPTION
**Pull request by silverwind**
Result of

```bash
perl -p -i -e 's#<textarea #<textarea dir="auto" #g' templates/**/* web_src/js/**/*.{js,vue}
perl -p -i -e 's#<input #<input dir="auto" #g' templates/**/* web_src/js/**/*.{js,vue}
````

There are likely a few superflous replacements, we need to review diff carefully.

Screenshot 2023-08-24 at 20 39 51
Screenshot 2023-08-24 at 20 46 13
Screenshot 2023-08-24 at 20 46 27

### CODE REVIEW

**Comment by silverwind** [comment]
I'm inclined to not fix the cases of hidden inputs or comments, just to reinforce that this attribute should always set.

BTW, this is a prime example why a `shared/input` template would be good. Change would be 1 line instead of 900+ :)

**Comment by yekanchi** [comment]
خیلی  ممنون.
کارتون عالیه!

File: web_src/js/modules/aria/checkbox.js [hunk 1]
```diff
@@ -23,8 +23,8 @@ function ariaCheckboxFn(...args) {
 }
 
 function attachInit(el) {
-  // Fomantic UI checkbox needs to be something like: <div class="ui checkbox"><label /><input /></div>
-  // It doesn't work well with <label><input />...</label>
+  // Fomantic UI checkbox needs to be something like: <div class="ui checkbox"><label /><input dir="auto" /></div>
+  // It doesn't work well with <label><input dir="auto" />...</label>
```
Comment (silverwind) [inline web_src/js/modules/aria/checkbox.js]:
```suggestion
  // Fomantic UI checkbox needs to be something like: <div class="ui checkbox"><label /><input/></div>
  // It doesn't work well with <label><input/>...</label>
```

File: templates/base/disable_form_autofill.tmpl [hunk 2]
```diff
@@ -4,7 +4,7 @@ Why we need to disable form autofill:
 2. We have `areYouSure` confirm dialog if a user leaves a pages without submit.
 Autofill will make the form changed even if the user didn't input anything. Then the user keeps seeing annoying confirm dialog.
 
-In history, Gitea put `<input class="fake" type="password">` in forms to bypass the autofill,
+In history, Gitea put `<input dir="auto" class="fake" type="password">` in forms to bypass the autofill,
```
Comment (silverwind) [inline templates/base/disable_form_autofill.tmpl]:
```suggestion
In history, Gitea put `<input class="fake" type="password">` in forms to bypass the autofill,
```

File: templates/shared/issuelist.tmpl [hunk 3]
```diff
@@ -4,7 +4,7 @@
 		<div class="flex-item flex-item-baseline">
 			<div class="flex-item-leading gt-ac">
 				{{if $.CanWriteIssuesOrPulls}}
-					<input type="checkbox" autocomplete="off" class="issue-checkbox gt-mr-4" data-issue-id={{.ID}} aria-label="{{$.locale.Tr "repo.issues.action_check"}} &quot;{{.Title}}&quot;">
+					<input dir="auto" type="checkbox" autocomplete="off" class="issue-checkbox gt-mr-4" data-issue-id={{.ID}} aria-label="{{$.locale.Tr "repo.issues.action_check"}} &quot;{{.Title}}&quot;">
```
Comment (silverwind) [inline templates/shared/issuelist.tmpl]: Doesn't really make sense on checkbox, but doesn't hurt either, should we we keep it?

File: templates/user/settings/profile.tmpl [hunk 4]
```diff
@@ -102,26 +102,26 @@
...
-						<input name="source" value="local" type="radio" {{if .SignedUser.UseCustomAvatar}}checked{{end}}>
+						<input dir="auto" name="source" value="local" type="radio" {{if .SignedUser.UseCustomAvatar}}checked{{end}}>
 						<label>{{.locale.Tr "settings.enable_custom_avatar"}}</label>
 					</div>
 				</div>
 
 				<div class="inline field gt-pl-4">
 					<label for="avatar">{{.locale.Tr "settings.choose_new_avatar"}}</label>
-					<input name="avatar" type="file" accept="image/png,image/jpeg,image/gif,image/webp">
+					<input dir="auto" name="avatar" type="file" accept="image/png,image/jpeg,image/gif,image/webp">
```
Comment (silverwind) [inline templates/user/settings/profile.tmpl]: Also unnecessary on file input.

**Comment by silverwind** [comment]
I think I'll actually do a pass and remove it from all inputs where type is hidden,checkbox,file. Any other types I missed?

**Comment by silverwind** [comment]
Actually it does work on `type=file`.

image

I think this is good to go. The extra work to remove it types `checkbox` and `hidden` seems not worth it.

**Review by yardenshoham — APPROVED** [review]
נראה מעולה, תודה רבה!

**Comment by wxiaoguang** [comment]
Haven't read code or thought about it carefully.

But I have a question about whether this approach is good enough: for future PRs, will there always be a person who are able to keep telling contributors "please add dir=auto" again and again?

**Comment by silverwind** [comment]
I think best we can do is watch out for these in review. MutationObserver might be something to consider too, yes.

**Comment by wxiaoguang** [comment]
> I think best we can do is watch out for these in review. MutationObserver might be something to consider too, yes.

Would you like to take a try? Or I could try too. TBH, keeping asking contributors to add "dir=auto" doesn't look good to me.

**Comment by silverwind** [comment]
Not interested right now, this PR should be a simple attribute addition. I will do one pass to remove it from `hidden` and `checkbox` just for the sake of it being useless there, but that's it.

**Comment by wxiaoguang** [comment]
> Not interested right now, this PR should be a simple attribute addition.

I will try to propose a MutationObserver solution. If there is no performance problem, I think that's the once and for all solution.

**Comment by silverwind** [comment]
Ready to review/merge now, diff is down to half after hidden/checkbox removal.

**Comment by wxiaoguang** [comment]
> Ready to review/merge now, diff is down to half after hidden/checkbox removal.

Could we wait for more time? I do not think keeping adding attributes manually is maintainable.

**Comment by silverwind** [comment]
> I will try to propose a MutationObserver solution. If there is no performance problem, I think that's the once and for all solution.

Try that I guess. Sadly it means LTR won't work without JS, but it would be the more maintainable solution via JS.

**Comment by wxiaoguang** [comment]
> Sadly it means RTL won't work without JS

Gitea UI already doesn't work without JS for long time.

**Comment by silverwind** [comment]
Well I blame it on the browser vendors for not making this attribute the default. Maybe they will see the light someday and we can remove the JS then.

**Comment by wxiaoguang** [comment]
->  Add "dir=auto" for input/textarea elements by default #26735

**Comment by silverwind** [comment]
https://github.com/go-gitea/gitea/pull/26735 has non-neglible performance impact, I think this solution is better. We don't have to meticulously have `dir=auto` on every element. For example GitHub only very rarely adds it, like on this Textarea.

File: web_src/js/modules/aria/checkbox.js [hunk 5]
```diff
@@ -23,8 +23,8 @@ function ariaCheckboxFn(...args) {
 }
 
 function attachInit(el) {
-  // Fomantic UI checkbox needs to be something like: <div class="ui checkbox"><label /><input /></div>
-  // It doesn't work well with <label><input />...</label>
+  // Fomantic UI checkbox needs to be something like: <div class="ui checkbox"><label /><input/></div>
+  // It doesn't work well with <label><input/>...</label>
```
Comment (silverwind) [inline web_src/js/modules/aria/checkbox.js]:
```suggestion
  // Fomantic UI checkbox needs to be something like: <div class="ui checkbox"><label /><input /></div>
  // It doesn't work well with <label><input />...</label>
```

**Comment by silverwind** [comment]
So I think I'm definitely done here, checked the diff twice, it's now down to  +399 −398.

**Comment by wxiaoguang** [comment]
#26735 has been fully rewritten, now it takes around 1.5ms on a PR page with +6966 -3615

**Comment by silverwind** [comment]
IDK, maybe we should just do it for the single markup textarea and be done with it, it seems that's exactly what GitHub is doing.

**Comment by silverwind** [comment]
Closing in favor of https://github.com/go-gitea/gitea/pull/26735

**Pull request closed without merging**